index.ts

```diff
@@ -450,34 +452,52 @@ export function trait<T extends SuiteTrait | TestTrait>(arg: T): T {
         return trait(function() {
+            // TODO: Implement an overload selector similar to the testOverload function
+
             if (arguments.length === 1) {
+                // Class decorator
                 const target = arguments[0];
                 target[traitSymbol] = value;
-            } else if (arguments.length === 2 && typeof arguments[1] === "string") {
+                return;
+            }
+
+            /* istanbul ignore if  */
+            if (arguments.length === 2 && typeof arguments[1] === "string") {
+                // PropertyDecorator, some TSC versions generated property decorators when decorating method
                 const target = arguments[0];
                 const property = arguments[1];
                 target[property][traitSymbol] = value;
-            } else if (arguments.length === 2) {
+                return;
+            }
+
+            if (arguments.length === 2) {
+                // Class trait as retries in `@suite(repeat(2)) class X {}`
                 const context: Mocha.ISuiteCallbackContext = arguments[0];
                 const ctor = arguments[1];
                 fn(context, value);
-            } else {
-                /* istanbul ignore else */
-                if (arguments.length === 3) {
-                    /* istanbul ignore else */
-                    if (typeof arguments[2] === "function") {
-                        const context: Mocha.ITestCallbackContext = arguments[0];
-                        const instance = arguments[1];
-                        const method = arguments[2];
-                        fn(context, value);
-                    } else if (typeof arguments[1] === "string") {
-                        const proto: Mocha.ITestCallbackContext = arguments[0];
-                        const prop = arguments[1];
-                        const descriptor = arguments[2];
-                        proto[prop][traitSymbol] = value;
-                    }
-                }
+                return;
+            }
+
+            if (arguments.length === 3 && typeof arguments[2] === "function") {
+                // Metod trait as retries in `@suite class { @test(retries(4)) method() {} }`
+                const context: Mocha.ITestCallbackContext = arguments[0];
+                const instance = arguments[1];
+                const method = arguments[2];
+                fn(context, value);
+                return;
+            }
+
+            /* istanbul ignore else */
+            if (arguments.length === 3 && typeof arguments[1] === "string") {
+                // MethodDecorator
+                const proto: Mocha.ITestCallbackContext = arguments[0];
+                const prop = arguments[1];
+                const descriptor = arguments[2];
+                proto[prop][traitSymbol] = value;
+                return;
             }
+
+            // assert unreachable.
```

Comment:
just remove

Reply:
I will replace this with proper assert.
Something like `assert.fail("Function called with incorrect arguments, expected ...");`